runtime/notification.go

```diff
@@ -0,0 +1,71 @@
+package runtime
+
+import (
+	"fmt"
+
+	"github.com/mitchellh/mapstructure"
+)
+
+type Notification struct {
+	GroupID   string      `json:"group_id"`
+	EventType string      `json:"event_type"`
+	Event     interface{} `json:"event"`
+}
+
+type StageEvent struct {
+	Name   string `json:"name"`
+	Action string `json:"action"` // entry or exit
+}
+
+type OutcomeEvent struct {
+	Outcome string `json:"outcome"` // ok
+}
+
+func NewStageNotification(groupId, name, action string) *Notification {
+	return &Notification{
+		GroupID:   groupId,
+		EventType: "stage",
+		Event: &StageEvent{
+			Name:   name,
+			Action: action,
+		},
+	}
+}
+
+func NewOutcomeOKNotification(groupId string) *Notification {
+	return &Notification{
+		GroupID:   groupId,
+		EventType: "outcome",
+		Event: &OutcomeEvent{
+			Outcome: "ok",
+		},
+	}
+}
+
+func (n *Notification) ParseEvent() error {
+	if n.EventType == "stage" {
+		r := &StageEvent{}
+		err := mapstructure.Decode(n.Event, r)
```

Comment:
Why do we need to use mapstructure here?

Reply:
Because we are consuming a `Notification`, but before unmarshaling it, we don't know what kind of notification. The kind is sent a `EventType` which is part of the notification.